lib/tesla/middleware/retry.ex

```diff
@@ -124,6 +143,19 @@ defmodule Tesla.Middleware.Retry do
     end
   end
 
+  defp should_retry_opt!(opts) do
+    case Keyword.get(opts, :should_retry, &match?({:error, _}, &1)) do
+      should_retry_fun when is_function(should_retry_fun, 1) ->
+        should_retry_fun
```

Comment:
I wonder if we should put a warning log to move away from 1 arity into 3 so there is only one encouraged way, and when breaking changes come, we could remove it.

Thoughts?!

Reply:
Pushing the user to transition to the 3-arity version could be quite beneficial since it would force them to think about the cases where this might be desirable. Having the option to keep it simple might be also nice, but imo hinting at customizing the retry logic is a good practice as in many situations you don't want to retry a POST request on a timeout for example.

If you think this is a good direction for the library, then I will add such a warning and modify the code accordingly :)

Reply:
Keep them both, we can come back to this later